### PR TITLE
Add ottl truncate editor

### LINKED ISSUE #23847: [processor/transform] not enough preconditions to guard against warnings
### Component(s)

processor/transform

### Is your feature request related to a problem? Please describe.

I have a transform processor configured to truncate log attributes and the body field:

```yaml
  transform/newrelic:
    log_statements:
      - context: resource
        statements:
          - truncate_all(attributes, 4090)
      - context: log
        statements:
          - truncate_all(attributes, 4090)
          - set(body, Substring(body.string, 0, 4090))
```
This works, but the problem is when we send a `body` whose length is less than 4090 this warning message occurs:
`invalid range for substring function, 4090 cannot be greater than the length of target string`

### Describe the solution you'd like

It would be nice to be able to have a precondition like `WHERE body.Length() > 4090` to be able to guard against this warning.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jack-berg**
**Description:** 

Add ottl `truncate` editor for truncating strings. E.g.:
```
    log_statements:
      - context: log
        statements:
        - truncate(body, 10)
```

**Link to tracking Issue:**  Resolves #23847.

**Testing:** Unit tests, local verification with running collector.

**Documentation:** Added documentation to `pkg/ottl/ottlfuncs/README.md`.

### CODE REVIEW

File: pkg/ottl/ottlfuncs/func_truncate.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type TruncateArguments[K any] struct {
+	Target ottl.GetSetter[K] `ottlarg:"0"`
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_truncate.go:14]: Let's use a StringGetter here to enforce the function only working on strings.

Comment (jack-berg) [reply pkg/ottl/ottlfuncs/func_truncate.go:14]: I did that initially but want to maintain the same update semantics of `update_all`. I didn't see a string specific option that allows for getting and setting. WDYT?

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_truncate.go:14]: You're right, I forgot this functions getter and setter are the same target.

File: pkg/ottl/ottlfuncs/func_truncate.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func truncate[K any](target ottl.GetSetter[K], limit int64) (ottl.ExprFunc[K], error) {
+	if limit < 0 {
+		return nil, fmt.Errorf("invalid limit for truncate function, %d cannot be negative", limit)
+	}
+	return func(ctx context.Context, tCtx K) (interface{}, error) {
+		if limit < 0 {
+			return nil, nil
+		}
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_truncate.go:39]: We can omit this duplicate check (truncate_all shouldn't have it either).

File: pkg/ottl/ottlfuncs/func_truncate.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	args, ok := oArgs.(*TruncateArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("TruncateFactory args must be of type *TruncateArguments[K]")
+	}
+
+	return truncate(args.Target, args.Limit)
+}
+
+func truncate[K any](target ottl.GetSetter[K], limit int64) (ottl.ExprFunc[K], error) {
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_truncate.go:32]: Can truncateAll be made to depend on this function?

**Review by TylerHelmuth — COMMENTED** [review]
Please add an `Enhancement` changelog entry for `pkg/ottl`.

File: pkg/ottl/ottlfuncs/func_truncate.go (hunk 3 above)
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_truncate.go:32]: Ya the `pcommon.Value` vs `ottl.GetSetter` ruins this idea.  Thanks for humoring me, but lets revert it.  I'm good keeping the change in truncate_all to remove the extra limit check tho.

File: .chloggen/add-ottl-truncate-editor.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/ottl
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add truncate editor
```
Comment (TylerHelmuth) [inline .chloggen/add-ottl-truncate-editor.yaml:12]:
```suggestion
note: Add new `truncate` editor that allows truncating a string to a specific length
```

**Review by TylerHelmuth — APPROVED** [review]

**Comment by jack-berg** [comment]
> Hmm, I know I'm swooping in late to this, but string editing is a bit more general than just truncate. Should we consider a general slice(start, end) function instead?

Sounds like competing concerns between consistency with existing `truncate_all` function and the increased flexibility of a more generic function. Could always start with `truncate` and add `slice` later. Adds a bit of bloat but maintaining these functions doesn't seem like too much overhead.

**Comment by jack-berg** [comment]
Bumping this. Wondering if any other @open-telemetry/collector-contrib-maintainer can review this and / or have opinions about a truncate vs. a more generic slice.

**Comment by TylerHelmuth** [comment]
I think @kentquirk brings up a good point that `truncate` might be too generic of a name for string-only truncation.  It is feasible that other types such as slices or time.Time may need truncated and if we reserve `truncate` for strings only that will make discoverability of the functions more difficult.

The easiest solution is to name it `truncate_str` so that the type is clear and we keep consistency with the word `truncate` and the parameter pattern used by `truncate_all` (which should've been called `truncate_all_strs`).

@kentquirk @evan-bradley what do you think?

**Comment by evan-bradley** [comment]
>I think @kentquirk brings up a good point that truncate might be too generic of a name for string-only truncation. It is feasible that other types such as slices or time.Time may need truncated and if we reserve truncate for strings only that will make discoverability of the functions more difficult.

I understood Kent's point a little differently: that `truncate` only trims the end of a string, as where we may also want to trim the beginning, which we currently support with the `Substring` function. What you said is also true though and I think is important to keep in mind.

Looking at the original issue, I think having a `Len` Converter that outputs the length of a value would be the best immediate solution. I'm hesitant to add a `truncate` editor since we can already accomplish the same with `set(body, Substring(body, 0, 10) where IsString(body)`. Adding a truncate editor would make this case just a little bit easier by making it `truncate(body, 10) where IsString(body)`, but I think I'd prefer for OTTL to generally have one right way of doing something.

As another way of solving the issue, should `Substring` return an error when the length exceeds the string length? Looking at a few other languages, JavaScript and Python both just return the full string, and Go only returns an error because it treats the string as a slice type. What is actionable by the user if they hit this case?

**Comment by jack-berg** [comment]
> set(body, Substring(body, 0, 10) where IsString(body)

Just to confirm, that's not a solution for the problem today without a `Len` function since `Substring` will error when the length is shorter than `10` (in this case).

> As another way of solving the issue, should Substring return an error when the length exceeds the string length?

If `Len` was added it could be actionable since the user would have a way of bounding the argument to always produce correct results. Its not currently actionable when you want to use it to perform truncation.

> but I think I'd prefer for OTTL to generally have one right way of doing something.

👍 

If folks agree on adding `Len`, I'm happy to open an alternative PR with that implementation.

**Comment by evan-bradley** [comment]
>Just to confirm, that's not a solution for the problem today without a Len function since Substring will error when the length is shorter than 10 (in this case).

Yes, that's a good point. The full correct statement would be `set(body, Substring(body, 0, 10) where IsString(body) and Len(body) >= 10`.

>If Len was added it could be actionable since the user would have a way of bounding the argument to always produce correct results. Its not currently actionable when you want to use it to perform truncation.

Could you elaborate on this? If `Substring("OpenTelemetry", 0, 100) == "OpenTelemetry"` and `Substring` doesn't throw any errors, I think it would be valid to say that `"OpenTelemetry"` has been truncated to be under 100 characters. For the more general case of string manipulation, it's not clear to me whether `Substring("OpenTelemetry", 4, 100) == "Telemetry"` is invalid output.

**Comment by jack-berg** [comment]
Closing in favor of #24420.

**Pull request closed without merging**